### ISSUE TITLE
Deserialization time on already typechecked file

### ISSUE THREAD

**Issue by JacquesCarette** [agda/agda#4600]
In the middle of doing other profiling, I happened to profile something that was already typechecked, i.e. all of the `.agdai` files were there already, nothing needed redone.  To be precise, in agda-categories, from `src/` I did
```
agda --no-main -v profile:15 Categories/Category/Core.agda >& Core.agda.prof4
```
and looking into that file, I see
```
Total                      3,468ms          
Miscellaneous                200ms          
Deserialization            2,838ms (3,099ms)
Deserialization.Compaction   260ms          
Import                       137ms          
Parsing                       30ms      
```
Why is there a full 3 seconds of deserialization?  Yes, it uses some things from the standard library, but not very much (or so I thought).

By comparison, if I hand-delete the corresponding `.agdai` from `_build/...`, I get
```
Checking Categories.Category.Core (/Users/carette/agda-categories/src/Categories/Category/Core.agda).
Total                        4,785ms          
Miscellaneous                  219ms          
Deserialization              3,092ms (3,369ms)
Deserialization.Compaction     276ms          
Serialization                  253ms   (406ms)
Serialization.BinaryEncode      82ms          
Serialization.BuildInterface    52ms          
Typing                         150ms   (365ms)
Typing.TypeSig                  94ms          
Typing.CheckRHS                 91ms          
Typing.CheckLHS                 11ms    (14ms)
Typing.OccursCheck              14ms          
Import                         191ms          
DeadCode                       105ms          
Parsing                         67ms   (104ms)
Parsing.OperatorsExpr           33ms          
Scoping                         22ms    (29ms)
Positivity                      24ms          
Accumulated statistics
A.Name  (fresh)       1,386
A.QName  (fresh)      1,164
Double  (fresh)          12
Integer  (fresh)          3
Node  (fresh)        32,690
Shared Term  (fresh)      0
String  (fresh)         545
Text  (fresh)             1
max-open-constraints      0
metas                   623
pointers  (fresh)         0
```
In other words, all the 'hard' work takes ~1s, but there's 3s of Deserialization overhead on top of that.

For files higher up in the hierarchy, it can take up to 15s to "do nothing".

Obviously there's something I am not understanding here - could someone please enlighten me?  And, more importantly, is there some way to cut out that time?  I was kind of hoping that typechecking an already checked file would cost essentially 0 time. [It can't be 0, a bunch of things on the filesystem have to be checked to make sure they are not dirty, but I was expecting this to be < 100ms on a modern computer with an SSD.]

This is agda-2.6.1 on MacOS X.

**Comment by nad** [comment]
Possibly related: #4465.

**Comment by UlfNorell** [comment]
The typical case when loading an interface is that you have some module that you are actually type checking that needs the contents of this interface. In this case I hope it's clear that deserialization (unpacking the stored interface into the proof state) takes a non-trivial amount of time.

I suppose it would be possible to optimize the case where *no* module needs type checking and do nothing in that case, but it would require a major overhaul of the import chasing logic. While such an overhaul wouldn't be unwelcome it hardly seems worth it for this particular issue.

**Comment by JacquesCarette** [comment]
Yes, it's clear in the typical case. [Although it seems that this is still over-eager, as loading a large module but with a very restrictive `using` doesn't seem to have any effect on load time].

I also agree that this use case is probably not worth optimizing for.  But what about the `using` cases? Is there an agda idiom that actually only loads exactly what I ask for, and nothing else?  I do realize that means worse error messages.  So it shouldn't be the default idiom. But for a large library, it would be nice if 'developer precision' translated to much faster load times.

**Comment by nad** [comment]
> Is there an agda idiom that actually only loads exactly what I ask for, and nothing else?

What about all the dependencies of what you ask for?

Agda currently has no support for loading parts of interface files, and the interface file format is not designed for this.

**Comment by JacquesCarette** [comment]
> What about all the dependencies of what you ask for?

Right, I should have been precise: what I ask for, and everything needed for that to make sense, i.e. all transitive dependencies. 

What this implies, for me, is that I should break the files in `agda-categories` into much smaller chunks, and then I'll have the option of being minimal. The "top level" files won't change in what they export, but what they'll become is a bunch of `open import XXX public`.  @MatthewDaggitt and @gallais should also be aware, as this might also be relevant for `agda-stdlib`. For `agda-categories`, this will likely speed up initial load of some files quite a bit.

**Comment by nad** [comment]
Given that the Emacs mode caches interfaces I'm not convinced that it makes sense to optimise things in this way. You might save tens of seconds in load time, but if a library becomes harder to navigate you might lose much more in navigation time, so perhaps it makes more sense to optimise for readability.

However, it might make sense to make the loading of interface files faster.

When I implemented compaction I noticed that the `compact` library makes it possible to serialise compact regions. I can imagine that reading regions serialised in this way would be faster than the current procedure, but there are some pretty serious [restrictions](https://hackage.haskell.org/package/compact-0.2.0.0/docs/Data-Compact-Serialize.html#v:hUnsafeGetCompact): "in practice, this means using static linking, using the exact same binary and turning off ASLR".

**Comment by MatthewDaggitt** [comment]
>  @MatthewDaggitt and @gallais should also be aware, as this might also be relevant for agda-stdlib. For agda-categories, this will likely speed up initial load of some files quite a bit.

Thanks for the ping. @gallais in particular is already aware of this, and has done some great work in ensuring that the contents of the `Base` modules are designed so that they have minimal dependencies and are snappy to load without dragging in all the proofs. Though there are still some sticking points (e.g. agda/agda-stdlib#1033)

**Comment by JacquesCarette** [comment]
After thinking about this for a while, my conclusion is:
1. right now, there is no "low hanging fruit" on the Agda side (thus closing)
2. for library builders [but not for everyone], a design consideration: make many small modules, so that 'higher up' modules can be more precise on their imports; also make giant convenience modules for end-users.

There is, of course, a blurry line between users and library developers. But I think most users will 'know' when they've crossed over.

I see that starting in 1.3 of stdlib, this has already started, and continuing with 1.4.  I'm starting the same process with agda-categories. It's tedious, but has already made some non-trivial difference on load times.
